### PR TITLE
PLG-733 Service API for product models scores

### PR DESCRIPTION
**Pull request by etiennekerbastard**
**Description (for Contributor and Core Developer)**

Setting up service API for product model + renaming of DQI files containing "Latest"

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/GetProductModelScoresQuery.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Akeneo PIM Enterprise Edition.
```
Comment (LaurentPetard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/GetProductModelScoresQuery.php:6]: Bad copyright

**Review by LaurentPetard — APPROVED** [review]

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Akeneo\Pim\Automation\DataQualityInsights\Domain\Query\ProductEvaluation;
+
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\Model\ChannelLocaleRateCollection;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface GetProductModelScoresByCodesQueryInterface
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php:13]: As we are using the notion of `byCode` filtering with the public methods of the interface, could we simplify the name with from `GetProductModelScoresByCodesQueryInterface` to `GetProductModelScoresQueryInterface`

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public function byProductModelCode(string $productModelCode): ChannelLocaleRateCollection;
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php:15]:
```suggestion
    public function byCode(string $productModelCode): ChannelLocaleRateCollection;
```

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    /**
+     * @return array<ChannelLocaleRateCollection>
+     */
+    public function byProductModelCodes(array $productModelCodes): array;
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php:20]:
```suggestion
    public function byCodes(array $productModelCodes): array;
```

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Model/ProductModelScore.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Automation\DataQualityInsights\PublicApi\Model;
+
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductModelScore
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Model/ProductModelScore.php:11]: The `ProductScore` and `ProductModelScore`  models have the exact same structure. We should use the same model.
Maybe rename it into `QualityScore` .... to discuss with the team

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php (hunk 2 above)
Comment (BryanS29-akeneo) [reply src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php:13]: Yes we can, we name it like this to be in coherence with what already exists Product side : https://github.com/akeneo/pim-community-dev/blob/bf3c28ed96b2e69b8c03a0f94af12a870690af77/src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetLatestProductScoresByIdentifiersQuery.php#L17

**Review by olivierpontier — APPROVED** [review]

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface GetProductModelScoresByCodesQueryInterface
+{
+    public function byProductModelCode(string $productModelCode): ChannelLocaleRateCollection;
+
+    /**
+     * @return array<ChannelLocaleRateCollection>
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/Query/ProductEvaluation/GetProductModelScoresByCodesQueryInterface.php:18]:
```suggestion
     * @return array<string, ChannelLocaleRateCollection>
```

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetProductModelScoresByCodesQuery.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public function byProductModelCode(string $productModelCode): ChannelLocaleRateCollection
+    {
+        $productModelScores = $this->byProductModelCodes([$productModelCode]);
+
+        return $productModelScores[$productModelCode] ?? new ChannelLocaleRateCollection();
+    }
+
+    public function byProductModelCodes(array $productModelCodes): array
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/GetProductModelScoresByCodesQuery.php:29]: In order to apply the static analysis on this implementation of the methode `byProductModelCodes` could you add this PHPDoc comment,please?

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/GetProductModelScoresQuery.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GetProductModelScoresQuery implements GetProductModelScoresQueryInterface
+{
+    public function __construct(private GetProductModelScoresByCodesQuery $getProductModelScoresByCodesQuery)
+    {
+    }
+
+    public function byProductModelCodes(array $productModelCodes): array
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/GetProductModelScoresQuery.php:23]:
```suggestion
    /**
     * {@inheritdoc}
     */
    public function byProductModelCodes(array $productModelCodes): array
```

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/GetProductScoresQuery.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GetProductScoresQuery implements GetProductScoresQueryInterface
+{
+    public function __construct(private GetProductScoresByIdentifiersQuery $getProductScoresByIdentifiersQuery)
+    {
+    }
+
+    public function byProductIdentifiers(array $productIdentifiers): array
```
Comment (olivierpontier) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/GetProductScoresQuery.php:23]:
```suggestion
    /**
     * {@inheritdoc}
     */
    public function byProductIdentifiers(array $productIdentifiers): array
```

**Review by olivierpontier — APPROVED** [review]

**Review by ghoussard — APPROVED** [review]

**Review by StevenVAIDIE — APPROVED** [review]

**Pull request merged**